### ISSUE TITLE
Review assertions vs exceptions in ghoul

### ISSUE THREAD

**Issue by emiax** [OpenSpace/Ghoul#7]
In `lua_helper.cpp`, there are assertions to check if the file exists. Would it make sense to throw an exception instead? This would be symmetric to things like` ProgramObject::Build`, that throw an exception if any file is missing.

Current code:
```
void loadDictionaryFromFile(const string& filename, ghoul::Dictionary& dictionary,
                                                                         lua_State* state)
{
    ghoul_assert(!filename.empty(), "Filename must not be empty");
    ghoul_assert(FileSys.fileExists(absPath(filename)), "Filename must be a file");
    ...
```

Also, more generally, what kind of policy should we have for exceptions vs assertions? I generally like:

> Assertions are for checking for errors in your code logic. Never, ever, use them to check something that you don't control. Remember, if an assertion fails it means that your code is wrong

(http://stackoverflow.com/questions/1957645/when-to-use-an-assertion-and-when-to-use-an-exception#comment29739993_1957662)

Thoughts on this?

**Comment by GPayne** [comment]
I agree with that policy quote for exceptions vs. assertions. I also like:

> Assertions should be used to check something that should never happen, while an exception should be used to check something that might happen

**Comment by alexanderbock** [comment]
Right now, I think the lua helper class should use an exception instead of an assertion...  in principle and in the long run though, I think they should go the other way of the `Build` method using assertions instead.

Regarding a policy, I see the use of `assert` as a check of preconditions and postconditions of a function and additionally a class invariant for methods (https://www.cs.umd.edu/class/fall2002/cmsc214/Projects/P1/proj1.contract.html).  So as long as the preconditions (checked with asserts) are specified in the Doxygen comment with `\pre` and they are checkable by the caller, i'm fine with that.  Only when things are not checkable by the caller, exceptions will be thrown.  I think that has two benefits; 1. if we use exceptions, the caller might check the value as well, which might be expensive to do; 2. specifying the preconditions as part of the interface might be possible in C++ at some point (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0380r1.pdf), which would shift the natural specification towards the assertions.  In general I agree with Emil quote that "Remember, if an assertion fails it means that your code is wrong". And definitely "Never, ever, use them to check something that you don't control".

In short: if something is checked with an assertion, it needs to be specified in the Doxygen using `\pre`, thus making it the callers responsibility to check the precondition.

**Comment by emiax** [comment]
That makes a lot of sense. Any preconditions must be documented. A check should not have to be performed multiple times (both by caller and callee), if it's possible to avoid. 

Regarding this specific case and other methods dealing with disk IO:
There could be failures in reading (and parsing) a file, even if it exists. To make it possible to recover from those, I think it makes sense to throw an exception. If there is the possibility of throwing an exception anyway, it could perhaps make sense to throw FileNotFound exceptions as well.

**Comment by alexanderbock** [comment]
It depends on how longterm we want to think.  Ultimately, it would probably be even better to have the function just use a stream and let the function doesn't care about whether it is actually a file or an in-memory stream.  That way, we could save ourselves some disk accesses.

And yes, failures in reading or parsing should **definitely** be an exception and not an assertion as the consumer of the function cannot check that in advance (or it would be too expensive as it would mean that the file would need to be parsed twice).

For now, I'm fine with a FileNotFound exception (which is already defined somewhere in Ghoul, if I remember correctly)

**Comment by emiax** [comment]
FileNotFoundError is thrown as of f1533fea495beb7194d6f2c4abb681d27a332431
